### PR TITLE
ghproxy: add instructions on how to use

### PR DESCRIPTION
**Pull request by nikhita**
I think the wording can definitely be improved. Suggestions welcome.

### CODE REVIEW

File: ghproxy/README.md [hunk 1]
```diff
@@ -3,4 +3,18 @@
...
 
-ghProxy is designed to reduce API token usage by allowing many components to share a single ghCache. Note that components must use the same API token to benefit from the cache and avoid clobbering existing cache entries for other tokens.
\ No newline at end of file
+ghProxy is designed to reduce API token usage by allowing many components to share a single ghCache. Note that components must use the same API token to benefit from the cache and avoid clobbering existing cache entries for other tokens.
+
+## How to use
+
+To use ghProxy for your prow instance,
+
+- Create a `StorageClass` using [`gce-ssd-retain_storageclass.yaml`](/prow/cluster/gce-ssd-retain_storageclass.yaml)
```
Comment (alvaroaleman) [inline ghproxy/README.md]: What is the benefit of the storageclass here?

File: ghproxy/README.md [hunk 2]
```diff
@@ -3,4 +3,18 @@
...
+if on GCE or specify your own storage class.
+- Create a `PersistentVolumeClaim` and `Deployment` using the config [here](prow/cluster/ghproxy_deployment.yaml).
+- Finally, create a [`Service`](/prow/cluster/ghproxy_service.yaml).
+- In the deployments for your prow components, use `--github-endpoint=http://ghproxy` as shown
+[here](https://github.com/kubernetes/test-infra/blob/6cd6c2332d493b4d141822cbaa3ffe5806f64825/prow/cluster/hook_deployment.yaml#L43).
+To make sure that the client is able to bypass cache if it is temporarily unavailable,
+you can additionally use `--github-endpoint=https://api.github.com`. The endpoints are used in the order
```
Comment (alvaroaleman) [inline ghproxy/README.md]: What about `[...] in the order they are mentioned, the first endpoint that does not return a connection error will be used`?

File: ghproxy/README.md (hunk 1 above)
Comment (ibrasho) [reply ghproxy/README.md]: It's used in the [PVC mentioned below](https://github.com/kubernetes/test-infra/blob/master/prow/cluster/ghproxy_deployment.yaml#L35)

Comment (alvaroaleman) [reply ghproxy/README.md]: Okay sorry, let me rephrase my question: What is the advantage of using a Storageclass with `reclaimPolicy: Retain` here?

Comment (stevekuznetsov) [reply ghproxy/README.md]: If you redeploy `ghproxy` and you end up reclaiming the data by a new deployment, we want the data to not have been deleted.

File: ghproxy/README.md [hunk 3]
```diff
@@ -3,4 +3,18 @@
 ghProxy is a reverse proxy HTTP cache optimized for use with the GitHub API (https://api.github.com).
 It is essentially just a reverse proxy wrapper around [ghCache](/ghproxy/ghcache) with some additional prometheus instrumentation logic to monitor disk usage and push metrics to a prometheus push gateway.
 
-ghProxy is designed to reduce API token usage by allowing many components to share a single ghCache. Note that components must use the same API token to benefit from the cache and avoid clobbering existing cache entries for other tokens.
\ No newline at end of file
+ghProxy is designed to reduce API token usage by allowing many components to share a single ghCache. Note that components must use the same API token to benefit from the cache and avoid clobbering existing cache entries for other tokens.
+
+## How to use
+
+To use ghProxy for your prow instance,
```
Comment (stevekuznetsov) [inline ghproxy/README.md]: Instead of words here, why not 

```
kubectl apply -f prow/cluster/gce-ssd-retain_storageclass.yaml
kubectl apply -f prow/cluster/ghproxy_deployment.yaml
kuebctl apply -f prow/cluster/ghproxy_service.yaml
```

File: ghproxy/README.md [hunk 4]
```diff
@@ -3,4 +3,18 @@
...
+
+- Create a `StorageClass` using [`gce-ssd-retain_storageclass.yaml`](/prow/cluster/gce-ssd-retain_storageclass.yaml)
+if on GCE or specify your own storage class.
+- Create a `PersistentVolumeClaim` and `Deployment` using the config [here](prow/cluster/ghproxy_deployment.yaml).
+- Finally, create a [`Service`](/prow/cluster/ghproxy_service.yaml).
+- In the deployments for your prow components, use `--github-endpoint=http://ghproxy` as shown
```
Comment (stevekuznetsov) [inline ghproxy/README.md]: This depends on the service name but should be fine to have this value here

File: ghproxy/README.md (hunk 2 above)
Comment (stevekuznetsov) [inline ghproxy/README.md]: Instead of "additionally use" maybe "also add the `--github-...` flag" or something

File: ghproxy/README.md (hunk 1 above)
Comment (ibrasho) [reply ghproxy/README.md]: Doesn't that require manual intervention to produce that effect? If the PVC was deleted, at least the `PersistentVolume.Spec.ClaimRef.UID` field [needs to be cleared]((https://github.com/kubernetes/kubernetes/issues/48609#issuecomment-314066616)) to make it's available again to be bound to the same claim.

AFAIK, `reclaimPolicy: retain` only means that the PV is not deleted when the PVC is deleted. If the PVC is recreated then a new PV is provisioned (and a new persistent desk on GCP). If the PVC is not deleted and only the deployment is recreated, then the PVC should be bound to the same PV regardless of the `reclaimPolicy` set.

File: ghproxy/README.md [hunk 5]
```diff
@@ -3,4 +3,18 @@
...
\ No newline at end of file
+ghProxy is designed to reduce API token usage by allowing many components to share a single ghCache. Note that components must use the same API token to benefit from the cache and avoid clobbering existing cache entries for other tokens.
+
+## How to use
+
+To use ghProxy for your prow instance,
+
+- Create a `StorageClass` using [`gce-ssd-retain_storageclass.yaml`](/prow/cluster/gce-ssd-retain_storageclass.yaml)
+if on GCE or specify your own storage class.
+- Create a `PersistentVolumeClaim` and `Deployment` using the config [here](prow/cluster/ghproxy_deployment.yaml).
```
Comment (cjwagner) [inline ghproxy/README.md]: Bad link
```suggestion
- Create a `PersistentVolumeClaim` and `Deployment` using the config [here](/prow/cluster/ghproxy_deployment.yaml).
```

File: ghproxy/README.md (hunk 4 above)
Comment (cjwagner) [inline ghproxy/README.md]: We should explicitly mention that this should be done for all prow components that use the main Prow bot token. (i.e. deployments that use the `oauth-token` secret).

File: ghproxy/README.md (hunk 1 above)
Comment (stevekuznetsov) [reply ghproxy/README.md]: Interesting. My understanding of PV binding is definitely shakey. @cjwagner thoughts?

Comment (cjwagner) [reply ghproxy/README.md]: Yeah that makes a lot of sense. I guess we could remove that and just keep the storage class for the SSD type?

**Comment by fejta** [comment]
This sounds cool. Any chance you'll get back to it?
